### ISSUE TITLE
Setting cookies on a subdomain called from the main domain

### ISSUE THREAD

**Issue by manuelonate** [vvo/iron-session#411]
First of all. Great library!

I am trying to authenticate a fully static next app (i.e. no /pages/api or any SSR) using another next app.

So the static website is at http://testdomain.com and the other on admin.testdomain.com.

I use the following options which I have seen recommended on other issues.

`export default function withSession(handler) {

  return withIronSession(handler, {
    password: process.env.IRON_PWD,
    cookieName: process.env.COOKIE_NAME,
    cookieOptions: {
      // the next line allows to use the session in non-https environments like
      // Next.js dev mode (http://localhost:3000)
      secure: process.env.NODE_ENV === 'production',
      domain: 'testdomain.com',
      maxAge: eval(process.env.COOKIE_MAX_AGE),
    },
  });`

And on the /api/login:

`
export default withSession(async (req, res) => {

if (Boolean(userid)) {
    const user = { user: userid };
    req.session.set('user', user);
    await req.session.save();
    res.json({ ok: true, user: userid });
  }`

My problem is that if I call /api/login from the auth server (admin.testdomain.com) a cookie for the whole domain is created, but if I call it from the static website (testdomain.com), the cookie is not created. I have tested on FIrefox and Chrome in Windows.

On both cases, I see on the headers returned by the server there is the exact same setcookie header:

`<html><body>

set-cookie | appsession=Fe26.2*1*8620153def2bcff56f07c804018dd0c56089e95dff67b4eb0c0da03934576dc8*d47ixmWp5NQ9pYjDr2t6xQ*82YdvvQtAgMoB-mituzzmIOAKvIV8RyddMDwvplqrFrQG8GDdNqtvUOt4BUTRCMFXbgr0gC235HFZONPiJ2gGA*1637254836905*333e3c1df6dc975940b25b9335b2a69a36721a154f9161ab227abcd4d8af0f54*kCViVUNQOaAvMwLBmg9JofUc_K60LoHo3jz3UGjKRT0;  Max-Age=7776000; Domain=testdomain.com; Path=/; HttpOnly; SameSite=Lax
-- | --

</body>
</html>`

**Comment by vvo** [comment]
Hey there, in your testing, are both domains (admin.testdomain.com and testdomain.com) served over https?

**Comment by vvo** [comment]
After reading a bit, setting a cookie from admin.domain.com/page with a domain of "domain.com" should work.
But setting a cookie from domain.com/page with a domain of "admin.domain.com" will not work.

In your case, both set-cookie seems to indicate domain=testdomain.com so it should definitely be available on both domains. Unless one is in HTTP and the other in https.

I would recommend you first try to set cookies without iron-session, and see how it goes, because your current issue is more probably related to set cookie and how it works rather than iron-session which seems to set the right cookie headers.

**Comment by manuelonate** [comment]
Vincent,

Thank you very much for your message. 

> After reading a bit, setting a cookie from admin.domain.com/page with a domain of "domain.com" should work.

Right. It doesn't. I noticed and moved the actual page that does the login to admin.domain.com and it works perfectly. The cookie for '.domain.com' is created.

> But setting a cookie from domain.com/page with a domain of "admin.domain.com" will not work.

I understand that. Fortunately, I don't think it is required.

After continuing with the tests, the biggest problem is that calls from domain.com to admin.domain.com/api do not include the cookie. Rather, the cookie is sent to all requests for assets on domain.com (which I really don't need since most of these files do not have to be protected).

> I would recommend you first try to set cookies without iron-session, and see how it goes, because your current issue is more probably related to set cookie and how it works rather than iron-session which seems to set the right cookie headers.

Your suggestion to test setting the cookies without using iron-session is very good. I should have thought of it. What I thought of was using client-side cookies on domain.com and sending the encrypted data on the api call, but that is a much worse idea.

Anyway, I have continued reading about cross-domain and SameSite cookie policies and it looks like it could become a minefield in the future, with Google trying to crush competitors in the ad-serving business. Therefore, I would have to monitor future changes to these policies and modify the system accordingly.

So, I have decided to implement the authentication system in a single app on a single domain. 

I consider this issue closed. Thank you very much.
